File: python/reverse_linked_list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        
        if head is None:
            return head
        
        prev = None
        curr = head
        nxt = head.next
        
        while curr.next:
            curr.next = prev
            prev = curr
            curr = nxt
            nxt = curr.next
            
            
        curr.next = prev
        return curr

File: javascript/LL/reverse_linked_list.js
// Given a singly linked list node, return its reverse.

// Bonus: Can you do this in \mathcal{O}(1)O(1) space?

/**
 * class LLNode {
 *   constructor(val, next=null) {
 *     this.val = val
 *     this.next = next
 *   }
 * }
 */
 class Solution {
    solve(node) {
        let prev = null, curr = node

        while (curr) {
            const next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        }

        return prev
    }
}
